### ISSUE TITLE
More tolerant to buffer overruns in responses

### ISSUE THREAD

**Issue by devlearner** [tatsuhiro-t/nghttp2#164]
Hi,

Saw that nghttp2 got neater with recent commits: some housekeepings and more robust HTTP semantics/messaging validations, which is great! (thanks, tatsuhiro!)

In another test with nghttpx, by chance I seem to find another corner case as I catch a crash if a (buggy) backend server returns a response which payload length is greater than that specified in `Content-Length:` header.

 I describe a bit of my test setup as follows:

```
# nghttpx in proxy mode
$ nghttpx -f 127.0.0.1,111 -b 127.0.0.1,333 --backend-no-tls --frontend-no-tls -L INFO

# prepare a (buggy) server that's going to send a malformed reply;
# (probably not the best one of its kind, but) I use socat (which's kind-of one of my favourites)
$ socat TCP4-LISTEN:333,bind=127.0.0.1 -

# send a request from a client, like curl, to nghttpx
$ curl http://127.0.0.1:111/1g
```
The backend socat will receive a request:
```
GET /1g HTTP/1.1
Host: 127.0.0.1:333
X-Forwarded-Proto: http
Via: 1.1 nghttpx


```
In a text editor, I prepare the following reply first, which I'll cut and paste into the socat terminal to be sent to nghttpx:
```
HTTP/1.1 200
Content-Length: 3

abcdef

```
(Since the terminal (socat) is essentially line-buffered, I've included an empty line at the end so it got sent. 
* Considering the trailing `\n`, the correct `Content-Length:` should be 7 [`abcdef\n`]. 
* I copy'n paste it into the terminal --- instead of typing it line-by-line --- so that the response fits into one buffer/ chunk.)

Here's the log after nghttpx receives the response (and subsequently crashed):
```
[INFO] shrpx_https_upstream.cc:631 [DOWNSTREAM:0x36e96cf0] HTTP response header completed
[INFO] shrpx_https_upstream.cc:816 [UPSTREAM:0x36e96f58] HTTP response headers
HTTP/1.1 200 OK
Content-Length: 3
Server: nghttpx nghttp2/0.7.5-DEV
Via: 1.1 nghttpx


[INFO] shrpx_https_upstream.cc:785 [DOWNSTREAM:0x36e96cf0] HTTP response completed
[INFO] shrpx_http_downstream_connection.cc:693 [DCONN:0x36e97250] nproc != nread
[INFO] shrpx_http_downstream_connection.cc:442 [DCONN:0x36e97250] Detaching from DOWNSTREAM:0x36e96cf0
[INFO] shrpx_client_handler.cc:581 [CLIENT_HANDLER:0x36e8ca50] Pooling downstream connection DCONN:0x36e97250
[INFO] shrpx_https_upstream.cc:551 [DCONN:0x36e97250] Network error/general error
      8 [main] nghttpx ... open_stackdumpfile: Dumping stack trace to nghttpx.exe.stackdump
[nghttpx exited]
```

N.B. If I retry the above setup with this response (I replaced line-breaks with `\r`):
```
HTTP/1.1 200\rContent-Length: 3\r\rabc\r
```
(due to the last `\r`, the correct `Content-Length:` should be 4)
the log looks like the following:
```
[INFO] shrpx_https_upstream.cc:631 [DOWNSTREAM:0x36e96d40] HTTP response header completed
[INFO] shrpx_https_upstream.cc:816 [UPSTREAM:0x36e96cf0] HTTP response headers
HTTP/1.1 200 OK
Content-Length: 3
Server: nghttpx nghttp2/0.7.5-DEV
Via: 1.1 nghttpx


[INFO] shrpx_https_upstream.cc:785 [DOWNSTREAM:0x36e96d40] HTTP response completed
[INFO] shrpx_http_downstream_connection.cc:702 [DCONN:0x36e96f88] HTTP parser failure: (HPE_CB_message_begin) the on_message_begin callback failed
[INFO] shrpx_http_downstream_connection.cc:442 [DCONN:0x36e96f88] Detaching from DOWNSTREAM:0x36e96d40
[INFO] shrpx_client_handler.cc:581 [CLIENT_HANDLER:0x36e8ca50] Pooling downstream connection DCONN:0x36e96f88
[INFO] shrpx_https_upstream.cc:551 [DCONN:0x36e96f88] Network error/general error
      7 [main] nghttpx 5956 cygwin_exception::open_stackdumpfile: Dumping stack trace to nghttpx.exe.stackdump
```

I happen to notice the recent commit "Validate HTTP semantics by default", which the libnghttp2 checks:
>For both [client and server]:
>...
>* Make sure that content-length matches the amount of DATA we
  received.
>
>If validation fails, RST_STREAM of type PROTOCOL_ERROR is issued.

Essentially (in this test setup) it's the backend server to be blamed for sending a malformed reply;
but perhaps it'll be nice if we could catch this exception for nghttpx as well 
(following the messaging check in libnghttp2) (so that it could fail safely without crashing).

Thanks!

**Comment by tatsuhiro-t** [comment]
Thanks always for the great bug report.  I pushed the fix via commit 7f9de00.
The "Validate HTTP semantics by default" is meant for HTTP/2 traffic only.
In this particular issue, this is HTTP/1 related.

**Comment by devlearner** [comment]
I see : ) (The "Validate HTTP semantics" for HTTP/2 is still great anyway)
Thanks for the quick fix!
